### PR TITLE
fixes #12847 - api for job descriptions

### PR DESCRIPTION
**Pull request by stbenjam**
~~Built on top of Daniels' rails4 PR since I've migrated my dev environment.  Will rebase once https://github.com/theforeman/foreman_remote_execution/pull/96 is merged~~

### CODE REVIEW

**Comment by ares** [comment]
@stbenjam could you please rebase?

File: app/models/job_invocation.rb [hunk 1]
```diff
@@ -132,7 +132,7 @@ def generate_description!
     description_format.scan(key_re) { |key| input_hash[key.first] }
     self.description = description_format
     input_hash.each do |k, v|
-      self.description.gsub!(Regexp.new("%\{#{k}\}"), v)
+      self.description.gsub!(Regexp.new("%\{#{k}\}"), v || '')
```
Comment (stbenjam) [inline app/models/job_invocation.rb]: `.gsub!` must be passed a string, and when working outside the UI (API, tests, etc), v can be nil.

**Comment by stbenjam** [comment]
[test]

File: app/controllers/api/v2/job_invocations_controller.rb [hunk 2]
```diff
@@ -44,8 +45,9 @@ def show
       def create
         composer = JobInvocationComposer.from_api_params(job_invocation_params)
         composer.save!
-        ForemanTasks.async_task(::Actions::RemoteExecution::RunHostsJob, composer.job_invocation)
         @job_invocation = composer.job_invocation
+        @job_invocation.generate_description! if @job_invocation.description.blank?
+        composer.triggering.trigger(::Actions::RemoteExecution::RunHostsJob, @job_invocation)
```
Comment (ares) [inline app/controllers/api/v2/job_invocations_controller.rb]: thanks

**Comment by ares** [comment]
Thanks for the API patch and other improvements it includes. Merging.

**Pull request merged**